File: decorator-pattern/python/src/ShapeDecorator.py
# -*- coding: utf-8 -*-
"""
@author: jarry
"""

from src.Shape import Shape

# 抽象装饰类，是否继承Shape可选


class ShapeDecorator(Shape):

    def __init__(self, decorated_shape):
        self.decorated_shape = decorated_shape

    def draw(self):
        self.decorated_shape.draw()

File: decorator-pattern/java/src/ShapeDecorator.java
package src;

// 抽象装饰类，是否实现Shape可选
public abstract class ShapeDecorator implements Shape {
// public abstract class ShapeDecorator {
   protected Shape decoratedShape;

   public ShapeDecorator(Shape decoratedShape) {
      this.decoratedShape = decoratedShape;
   }

   public void draw() {
      decoratedShape.draw();
   }
}